README.rst
=====
PySia
=====


.. image:: https://img.shields.io/pypi/v/pysia.svg
        :target: https://pypi.python.org/pypi/pysia

.. image:: https://img.shields.io/travis/jnmclarty/pysia.svg
        :target: https://travis-ci.org/jnmclarty/pysia

.. image:: https://readthedocs.org/projects/pysia/badge/?version=latest
        :target: https://pysia.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status

.. image:: https://pyup.io/repos/github/jnmclarty/pysia/shield.svg
     :target: https://pyup.io/repos/github/jnmclarty/pysia/
     :alt: Updates


Sia API bindings for Python 2 & 3.

This library is built using code-generation, to exactly match the endpoints, docs and responses maintained by Sia.  This library will not attempt to improve on, or resist, any API changes made upstream to siad.

This version targets siad >= 1.2.2.

Install
-------

.. code-block:: console

    pip install pysia

Usage
-----

.. code-block:: python

   >>> from pysia import Sia
   >>> sc = Sia() # Optionally, pass host & port.  Defaults to localhost & 9980
   
   >>> consensus = sc.get_consensus()
   >>> consensus['height']
   108058
       
   backup_made = sc.get_wallet_backup(destination=r'd:\siadwallet.dat')
   print(backup_made)
   # True
   
   backup_made = sc.get_wallet_backup(destination=r'error causing input?@#$!`')
   print(backup_made)
   >>> {'message': 'error when calling /wallet/backup: destination must be an absolute path'}

   print(sc.get_gateway())
   >>> {'peers': [{'netaddress': '92.253.172.90:9981', 'version': '0.5.2', 'inbound': False, 'local': False},...]}

   >>> print(sc.set_gateway_connect('212.77.177.47:9981'))
   True

   >>> print(sc.set_gateway_disconnect('212.77.177.47:9981'))
   True
    
   >>> print(sc.set_gateway_disconnect('212.77.177.47:9981'))
   {'message': 'not connected to that node'}


Features
--------

* Exposes a method for each API endpoint of siad, matching Siad API docs 1-to-1
* User-friendly autocomplete (GET -> getters, POST -> setters)
* Pure python responses
* User-friendly key-words called out for url-parameters GET-methods

Donations
---------

Sia:

2fd5ada234b5dba82584160213d8c9698d080bc4311277667a1ef38e5265fe7058aeeb627822

License
-------

* Free software: MIT license

Documentation
-------------

Coming soon, here:

* Documentation: https://pysia.readthedocs.io.

See Also
--------

There are python 3 bindings for Sia, maintained by humans which might be or become more pythonic. See siapy_.

Credits
-------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage
.. _siapy: https://github.com/lolsteve/siapy


HISTORY.rst
=======
History
=======

0.1.122.0 (2017-05-28)
----------------------

* First code-gen produced version, targetting python 2 & 3 & siad 1.2.2


tests/test_pysia.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_pysia
----------------------------------

Tests for `pysia` module.
"""

import pytest

@pytest.fixture
def sia():
    from pysia import Sia
    return Sia()

def test_basic_methods_exist(sia):
    assert (hasattr(sia, 'get_consensus')), "A sia client should have a get_consensus method."


travis_pypi_setup.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Update encrypted deploy password in Travis config file
"""


from __future__ import print_function
import base64
import json
import os
from getpass import getpass
import yaml
from cryptography.hazmat.primitives.serialization import load_pem_public_key
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15


try:
    from urllib import urlopen
except:
    from urllib.request import urlopen


GITHUB_REPO = 'jnmclarty/pysia'
TRAVIS_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), '.travis.yml')


def load_key(pubkey):
    """Load public RSA key, with work-around for keys using
    incorrect header/footer format.

    Read more about RSA encryption with cryptography:
    https://cryptography.io/latest/hazmat/primitives/asymmetric/rsa/
    """
    try:
        return load_pem_public_key(pubkey.encode(), default_backend())
    except ValueError:
        # workaround for https://github.com/travis-ci/travis-api/issues/196
        pubkey = pubkey.replace('BEGIN RSA', 'BEGIN').replace('END RSA', 'END')
        return load_pem_public_key(pubkey.encode(), default_backend())


def encrypt(pubkey, password):
    """Encrypt password using given RSA public key and encode it with base64.

    The encrypted password can only be decrypted by someone with the
    private key (in this case, only Travis).
    """
    key = load_key(pubkey)
    encrypted_password = key.encrypt(password, PKCS1v15())
    return base64.b64encode(encrypted_password)


def fetch_public_key(repo):
    """Download RSA public key Travis will use for this repo.

    Travis API docs: http://docs.travis-ci.com/api/#repository-keys
    """
    keyurl = 'https://api.travis-ci.org/repos/{0}/key'.format(repo)
    data = json.loads(urlopen(keyurl).read().decode())
    if 'key' not in data:
        errmsg = "Could not find public key for repo: {}.\n".format(repo)
        errmsg += "Have you already added your GitHub repo to Travis?"
        raise ValueError(errmsg)
    return data['key']


def prepend_line(filepath, line):
    """Rewrite a file adding a line to its beginning.
    """
    with open(filepath) as f:
        lines = f.readlines()

    lines.insert(0, line)

    with open(filepath, 'w') as f:
        f.writelines(lines)


def load_yaml_config(filepath):
    with open(filepath) as f:
        return yaml.load(f)


def save_yaml_config(filepath, config):
    with open(filepath, 'w') as f:
        yaml.dump(config, f, default_flow_style=False)


def update_travis_deploy_password(encrypted_password):
    """Update the deploy section of the .travis.yml file
    to use the given encrypted password.
    """
    config = load_yaml_config(TRAVIS_CONFIG_FILE)

    config['deploy']['password'] = dict(secure=encrypted_password)

    save_yaml_config(TRAVIS_CONFIG_FILE, config)

    line = ('# This file was autogenerated and will overwrite'
            ' each time you run travis_pypi_setup.py\n')
    prepend_line(TRAVIS_CONFIG_FILE, line)


def main(args):
    public_key = fetch_public_key(args.repo)
    password = args.password or getpass('PyPI password: ')
    update_travis_deploy_password(encrypt(public_key, password.encode()))
    print("Wrote encrypted password to .travis.yml -- you're ready to deploy")


if '__main__' == __name__:
    import argparse
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', default=GITHUB_REPO,
                        help='GitHub repo (default: %s)' % GITHUB_REPO)
    parser.add_argument('--password',
                        help='PyPI password (will prompt if not provided)')

    args = parser.parse_args()
    main(args)


pysia/client.py
# -*- coding: utf-8 -*-

import requests as r
from simplejson import JSONDecodeError

GET = 1
POST = 2

class Sia(object):
    def __init__(self, host='http://localhost', port='9980'):
        self.host = host
        self.port = port
    @property
    def _url_base(self):
        return self.host + ":" + str(self.port)
    def __call__(self, verb, url, data=None):
        user_agent = {'User-agent': 'Sia-Agent'}
        full_url = self._url_base + url
        if verb == GET:
            resp = r.get(full_url, headers=user_agent, params=data)
        elif verb == POST:
            resp = r.post(full_url, headers=user_agent, data=data)
        try:
            return resp.json()
        except JSONDecodeError:
            return resp.ok

    def get_consensus(self, **data):
        url = '/consensus'
        return self(GET, url, data)

    def set_consensus_validate_transactionset(self, **data):
        url = '/consensus/validate/transactionset'
        return self(POST, url, data)

    def get_daemon_constants(self, **data):
        url = '/daemon/constants'
        return self(GET, url, data)

    def get_daemon_stop(self, **data):
        url = '/daemon/stop'
        return self(GET, url, data)

    def get_daemon_version(self, **data):
        url = '/daemon/version'
        return self(GET, url, data)

    def get_gateway(self, **data):
        url = '/gateway'
        return self(GET, url, data)

    def set_gateway_connect(self, netaddress, **data):
        url = '/gateway/connect/{netaddress}'.format(netaddress=netaddress)
        return self(POST, url, data)

    def set_gateway_disconnect(self, netaddress, **data):
        url = '/gateway/disconnect/{netaddress}'.format(netaddress=netaddress)
        return self(POST, url, data)

    def get_host(self, **data):
        url = '/host'
        return self(GET, url, data)

    def set_host(self, **data):
        url = '/host'
        return self(POST, url, data)

    def set_host_announce(self, **data):
        url = '/host/announce'
        return self(POST, url, data)

    def get_host_storage(self, **data):
        url = '/host/storage'
        return self(GET, url, data)

    def set_host_storage_folders_add(self, **data):
        url = '/host/storage/folders/add'
        return self(POST, url, data)

    def set_host_storage_folders_remove(self, **data):
        url = '/host/storage/folders/remove'
        return self(POST, url, data)

    def set_host_storage_folders_resize(self, **data):
        url = '/host/storage/folders/resize'
        return self(POST, url, data)

    def set_host_storage_sectors_delete(self, merkleroot, **data):
        url = '/host/storage/sectors/delete/{merkleroot}'.format(merkleroot=merkleroot)
        return self(POST, url, data)

    def get_hostdb_active(self, **data):
        url = '/hostdb/active'
        return self(GET, url, data)

    def get_hostdb_all(self, **data):
        url = '/hostdb/all'
        return self(GET, url, data)

    def get_hostdb_hosts(self, pubkey, **data):
        url = '/hostdb/hosts/{pubkey}'.format(pubkey=pubkey)
        return self(GET, url, data)

    def get_miner(self, **data):
        url = '/miner'
        return self(GET, url, data)

    def get_miner_header(self, **data):
        url = '/miner/header'
        return self(GET, url, data)

    def set_miner_header(self, **data):
        url = '/miner/header'
        return self(POST, url, data)

    def get_miner_start(self, **data):
        url = '/miner/start'
        return self(GET, url, data)

    def get_miner_stop(self, **data):
        url = '/miner/stop'
        return self(GET, url, data)

    def get_renter(self, **data):
        url = '/renter'
        return self(GET, url, data)

    def set_renter(self, **data):
        url = '/renter'
        return self(POST, url, data)

    def get_renter_contracts(self, **data):
        url = '/renter/contracts'
        return self(GET, url, data)

    def set_renter_delete(self, siapath, **data):
        url = '/renter/delete/{siapath}'.format(siapath=siapath)
        return self(POST, url, data)

    def get_renter_download(self, siapath, **data):
        url = '/renter/download/{siapath}'.format(siapath=siapath)
        return self(GET, url, data)

    def get_renter_downloadasync(self, siapath, **data):
        url = '/renter/downloadasync/{siapath}'.format(siapath=siapath)
        return self(GET, url, data)

    def get_renter_downloads(self, **data):
        url = '/renter/downloads'
        return self(GET, url, data)

    def get_renter_files(self, **data):
        url = '/renter/files'
        return self(GET, url, data)

    def get_renter_prices(self, **data):
        url = '/renter/prices'
        return self(GET, url, data)

    def set_renter_rename(self, siapath, **data):
        url = '/renter/rename/{siapath}'.format(siapath=siapath)
        return self(POST, url, data)

    def set_renter_upload(self, siapath, **data):
        url = '/renter/upload/{siapath}'.format(siapath=siapath)
        return self(POST, url, data)

    def get_wallet(self, **data):
        url = '/wallet'
        return self(GET, url, data)

    def set_wallet_033x(self, **data):
        url = '/wallet/033x'
        return self(POST, url, data)

    def get_wallet_address(self, **data):
        url = '/wallet/address'
        return self(GET, url, data)

    def get_wallet_addresses(self, **data):
        url = '/wallet/addresses'
        return self(GET, url, data)

    def get_wallet_backup(self, **data):
        url = '/wallet/backup'
        return self(GET, url, data)

    def set_wallet_init(self, **data):
        url = '/wallet/init'
        return self(POST, url, data)

    def set_wallet_init_seed(self, **data):
        url = '/wallet/init/seed'
        return self(POST, url, data)

    def set_wallet_lock(self, **data):
        url = '/wallet/lock'
        return self(POST, url, data)

    def set_wallet_seed(self, **data):
        url = '/wallet/seed'
        return self(POST, url, data)

    def get_wallet_seeds(self, **data):
        url = '/wallet/seeds'
        return self(GET, url, data)

    def set_wallet_siacoins(self, **data):
        url = '/wallet/siacoins'
        return self(POST, url, data)

    def set_wallet_siafunds(self, **data):
        url = '/wallet/siafunds'
        return self(POST, url, data)

    def set_wallet_siagkey(self, **data):
        url = '/wallet/siagkey'
        return self(POST, url, data)

    def set_wallet_sweep_seed(self, **data):
        url = '/wallet/sweep/seed'
        return self(POST, url, data)

    def get_wallet_transaction(self, id, **data):
        url = '/wallet/transaction/{id}'.format(id=id)
        return self(GET, url, data)

    def get_wallet_transactions(self, addr=None, **data):
        if addr:
            url = '/wallet/transactions/{addr}'.format(addr=addr)
        else:
            url = '/wallet/transactions'
        return self(GET, url, data)

    def set_wallet_unlock(self, **data):
        url = '/wallet/unlock'
        return self(POST, url, data)


if __name__ == '__main__':
    sc = Sia()
    cs = sc.get_consensus()
    print(cs['height'])
    # 108060
    
    backup_made = sc.get_wallet_backup(destination=r'd:\siadwallet.dat')
    print(backup_made)
    # True
    
    backup_made = sc.get_wallet_backup(destination=r'error causing input?@#$!`')
    print(backup_made)
    # {'message': 'error when calling /wallet/backup: destination must be an absolute path'}

    print(sc.get_gateway())
    # {'peers': [{'netaddress': '92.253.172.90:9981', 'version': '0.5.2', 'inbound': False, 'local': False}, {'netaddress': '176.9.43.109:9981', 'version': '1.1.2', 'inbound': False, 'local': False}, {'netaddress': '91.134.136.124:9981', 'version': '1.2.0', 'inbound': False, 'local': False}, {'netaddress': '76.190.165.207:9981', 'version': '1.2.1', 'inbound': False, 'local': False}, {'netaddress': '51.15.58.86:9981', 'version': '1.1.2', 'inbound': False, 'local': False}, {'netaddress': '37.139.15.138:9981', 'version': '1.0.0', 'inbound': False, 'local': False}, {'netaddress': '87.98.189.200:9981', 'version': '1.2.2', 'inbound': False, 'local': False}], 'netaddress': '99.244.212.203:9981'}

    print(sc.set_gateway_connect('212.77.177.47:9981'))
    # True

    print(sc.set_gateway_disconnect('212.77.177.47:9981'))
    # True
    
    print(sc.set_gateway_disconnect('212.77.177.47:9981'))
    # {'message': 'not connected to that node'}
